src/deploy/extensions/prepare.ts

```diff
@@ -15,37 +15,53 @@ import { displayWarningsForDeploy, outOfBandChangesWarning } from "../../extensi
 import { detectEtagChanges } from "../../extensions/etags";
 import { checkSpecForV2Functions, ensureNecessaryV2ApisAndRoles } from "./v2FunctionHelper";
 import { acceptLatestAppDeveloperTOS } from "../../extensions/tos";
+import {
+  extractAllDynamicExtensions,
+  extractExtensionsFromBuilds,
+} from "../../extensions/runtimes/common";
+import { Build } from "../functions/build";
+import { normalizeAndValidate } from "../../functions/projectConfig";
+import { getEndpointFilters, targetCodebases } from "../functions/functionsDeployHelper";
 
-export async function prepare(context: Context, options: Options, payload: Payload) {
-  context.extensionsStartTime = Date.now();
+// This is called by prepare and also prepareDynamicExtensions. The only difference
+// is which set of extensions is in the want list and which is in the noDelete list.
+// isPrimaryCall is true exactly once per deploy. So if you have just 'firebase deploy'
+// it will be true when called from extensions/prepare but false when called from
+// functions/prepare. If you have 'firebase deploy --only functions' then it will
+// be true when called from functions/prepare (since extensions/prepare would
+// not be called). It is necessary otherwise you can get the same questions
+// and notifications twice (e.g. delete these extensions?)
+async function prepareHelper(
+  context: Context,
+  options: Options,
+  payload: Payload,
+  wantExtensions: planner.DeploymentInstanceSpec[],
+  noDeleteExtensions: planner.DeploymentInstanceSpec[],
+  isPrimaryCall: boolean,
```

Comment:
`isPrimaryCall` strikes me as a code smell here - it seems like it may be a sign that we ought to split out the prompting code from the rest of the prepare code here.

If you want to avoid refactors here, maybe rename `isPrimaryCall` to `shouldDeleteExtensions` - AFAICT, that is the only behavior we actually gate on this.

Reply:
Renamed.